extensions/wikia/IndexingPipeline/PipelineEventProducer.class.php

```diff
@@ -1,69 +1,173 @@
+		$ns = self::getArticleNamespace( $article->getTitle() );
+		$action = $rev->getPrevious() === null ? self::ACTION_CREATE : self::ACTION_UPDATE;
+
+		self::send( 'onNewRevisionFromEditComplete', $article->getId() );
+		self::nSend( $action, $article->getId(), $ns );
+
 		return true;
 	}
 
+	/**
+	 * @desc Fires on:
+	 *  - article delete
+	 * @return bool
+	 */
 	static public function onArticleDeleteComplete( &$oPage, &$oUser, $reason, $pageId ) {
-		wfDebug( "IndexingPipeline:onArticleDeleteComplete\n" );
+		$ns = self::getArticleNamespace( $oPage->getTitle() );
+
 		self::send( 'onArticleDeleteComplete', $pageId );
+		self::nSend( self::ACTION_DELETE, $pageId, $ns );
+
 		return true;
 	}
 
+	/**
+	 * @desc Fires on:
+	 *  - article restore
+	 * Send ACTION_CREATE as an article with new ID is created
+	 * @return bool
+	 */
 	static public function onArticleUndelete( Title &$oTitle, $isNew = false ) {
-		wfDebug( "IndexingPipeline:onArticleUndelete\n" );
+		$ns = self::getArticleNamespace( $oTitle );
+
 		self::send( 'onArticleUndelete', $oTitle->getArticleId(), [ 'isNew' => $isNew ] );
+		self::nSend( self::ACTION_CREATE, $oTitle->getArticleId(), $ns, [ 'isNew' => $isNew ] );
+
 		return true;
 	}
 
+	/**
+	 * @desc Fires on:
+	 *  - article rename
+	 * Send ACTION_UPDATE as the ID of article remains the same only title changes
+	 * @return bool
+	 */
 	static public function onTitleMoveComplete( &$oOldTitle, &$oNewTitle, &$oUser, $pageId, $redirectId = 0 ) {
-		wfDebug( "IndexingPipeline:onTitleMoveComplete\n" );
+		$ns = self::getArticleNamespace( $oNewTitle );
+
 		self::send( 'onTitleMoveComplete', $pageId, [ 'redirectId' => $redirectId ] );
+		self::nSend( self::ACTION_UPDATE, $pageId, $ns, [ 'redirectId' => $redirectId ] );
+
 		return true;
 	}
+
+	/*
+	 * Helper methods
+	 */
+
+	/**
+	 * @desc create message with cityId and pageId fields
+	 * @param $pageId
+	 * @return \stdClass
+	 */
+	protected static function prepareMessage( $pageId ) {
+		global $wgCityId;
+		$msg = new stdClass();
+		$msg->cityId = $wgCityId;
+		$msg->pageId = $pageId;
+
+		return $msg;
+	}
+
+	/**
+	 * @param $key
+	 * @param $data
+	 */
+	protected static function publish( $key, $data ) {
+		try {
+			self::getPipeline()->publish( $key, $data );
+		} catch ( Exception $e ) {
+			\Wikia\Logger\WikiaLogger::instance()->error( $e->getMessage() );
+		}
+	}
+
+	/** @return PipelineConnectionBase */
+	protected static function getPipeline() {
+		if ( !isset( self::$pipe ) ) {
+			self::$pipe = new PipelineConnectionBase();
+		}
+
+		return self::$pipe;
+	}
+
+	protected static function getArticleNamespace( $title ) {
+		global $wgCityId;
+
+		$pageNamespace = $title->getNamespace();
+		$contentNamespaces = WikiFactory::getVarValueByName( self::WG_CONTENT_NAMESPACES_KEY, $wgCityId );
+		if ( in_array($pageNamespace, $contentNamespaces) ) {
+			$pageNamespace =  self::CONTENT;
+		} else {
+			$pageNamespace = $title->getNsText();
```

Comment:
What are the possible characters in getNsText()'s output? Is there a need to sanitize it?